### PR TITLE
Allow valid zero tweaks

### PR DESCRIPTION
**Pull request by starius**
## Summary

This PR allows aggregate-key tweaking to represent and apply a zero-valued tweak while still rejecting tweaks outside BIP327's valid scalar range. It also updates the tweak-related test loaders so future BIP327 vectors can exercise zero tweaks instead of rejecting them during test parsing.

## Motivation

The previous API could not apply a tweak value of zero because it required a non-zero scalar representation. That rejected a value that BIP327 permits. For plain tweaks, zero is a no-op, but for x-only tweaks a zero tweak can still be meaningful because the x-only rule may normalize an odd aggregate key to its even-Y representative and update the accumulated sign and tweak state.

## Compatibility

This changes the public tweak-method bound from `Into<Scalar>` to `Into<MaybeScalar>` so zero can be represented directly. Existing callers that pass the crate's `Scalar`, `MaybeScalar`, 32-byte tweak values, or supported backend scalar/key types remain covered. Callers with custom wrapper types that only implemented conversion into the crate's non-zero `Scalar` may need to implement conversion into `MaybeScalar` or pass a supported tweak representation directly.

## BIP327 clarification

[BIP327's overview of aggregate-key tweaking](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#tweaking-the-aggregate-public-key) says that `ApplyTweak` takes an `is_xonly_t` mode, with `false` for plain tweaking and `true` for x-only tweaking. The [KeyAgg Context](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#keyagg-context) stores `(Q, gacc, tacc)`, where `Q` is the potentially tweaked aggregate public key, `tacc` is the accumulated tweak, and `gacc` tracks accumulated sign flips.

The relevant rule is in [BIP327 `ApplyTweak`](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#applying-tweaks): if `is_xonly_t` is true and `Q` does not have even Y, set `g = -1 mod n`; otherwise set `g = 1`. Then parse `t = int(tweak)`, failing only if `t >= n`, and compute `Q' = g*Q + t*G`, `gacc' = g*gacc mod n`, and `tacc' = t + g*tacc mod n`.

So if `t = 0`, a plain tweak gives `Q' = Q` and `tacc' = tacc`; it is just a no-op. But an x-only zero tweak is not always a no-op. If the current `Q` has odd Y, BIP327 sets `g = -1`, so `Q' = -Q`, `gacc' = -gacc mod n`, and `tacc' = -tacc mod n`. The x-coordinate is unchanged, but the plain point has been changed from the odd-Y representative to the even-Y representative, and the parity accumulator has changed.

This also matches the [BIP327 tweaking definition](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#tweaking-definition), which describes x-only tweaking as using `with_even_y(Q_i-1)` when `is_xonly_t_i` is true, and later states that the sign factor `g_i-1` is `-1 mod n` exactly when the tweak is x-only and the previous aggregate point has odd Y.

That is why zero needs to be representable by the tweak API. BIP327 accepts tweak integers in the range `0 <= t < n`; zero is valid. For plain tweaks it is only a no-op, but for x-only tweaks it can still perform the normalization step and update `gacc`/`tacc` as specified.

## Tests

This PR adds regression tests for a plain zero tweak, x-only zero tweaks applied to odd aggregate keys with and without an existing parity accumulator, and serialization of the newly reachable state where the parity accumulator is set while the tweak accumulator is zero. The x-only tests cover the cases where the x-coordinate is unchanged but the aggregate point parity and accumulated tweak state must still be updated.

### CODE REVIEW

**Comment by conduition** [comment]
Thanks for this @starius. I wanted to get #17 out ASAP so I already tagged & released those changes.  I'll try to review this one early next week.

File: src/key_agg.rs [hunk 1]
```diff
@@ -209,7 +275,7 @@ impl KeyAggContext {
     pub fn with_tweaks<S, I>(mut self, tweaks: I) -> Result<Self, TweakError>
     where
         I: IntoIterator<Item = (S, bool)>,
-        S: Into<Scalar>,
+        S: IntoTweak,
```
Comment (conduition) [inline src/key_agg.rs]: Rather than introducing a new trait, I think I might prefer simply switching to `impl Into<MaybeScalar>` for the tweak parameter's type. This would be backwards compatible with anyone passing `Scalar`, or most types that implement `Into<Scalar>`, like the various backbone secret key types (`secp256k1::SecretKey`, `k256::NonZeroScalar`, etc).

This also removes the error condition because it's impossible to pass a scalar which greater than or equal to `n`.

**Review by conduition — CHANGES_REQUESTED** [review]

File: src/key_agg.rs (hunk 1 above)
Comment (starius) [reply src/key_agg.rs]: I tried to get rid of `IntoTweak` in favor of `Into<MaybeScalar>`:
diff: https://gist.github.com/starius/24c9e0a4c311a5d25f5d2d17eef65e66

It mostly works, but panics on `secp256k1::Scalar::ZERO`:
log: https://gist.github.com/starius/5b6742c4296c28cff8a0896ef4915b74

The reason of the panic is that in secp 0.6.0, `From<secp256k1::Scalar>` for `MaybeScalar` routes through non-zero `secp::Scalar`, so zero panics instead of becoming `MaybeScalar::Zero`. My original version (IntoTweak) avoids that by parsing `secp256k1::Scalar::ZERO.to_be_bytes()` through `MaybeScalar::try_from`.

The panic was fixed when I replaced `secp256k1::Scalar::ZERO` with `MaybeScalar::Zero` in `.with_plain_tweak()` and `.with_xonly_tweak` calls in tests.

We can keep it like that and document `with_xonly_tweak`, `with_plain_tweak` to pass `MaybeScalar::Zero` not `secp256k1::Scalar::ZERO`.

What do you think?

Comment (conduition) [reply src/key_agg.rs]: > The reason of the panic is that in secp 0.6.0, `From<secp256k1::Scalar>` for `MaybeScalar` routes through non-zero `secp::Scalar`, so zero panics instead of becoming `MaybeScalar::Zero`

Hmm that sounds like something I need to fix in `secp`. I'll see about patching that bug, and once i've published a new version we can have another go

Comment (conduition) [reply src/key_agg.rs]: The upstream panic in `secp` should be fixed now with v0.7.0 https://github.com/conduition/secp/commit/84a9c4c1d23acc328a300270dcccbdc191b383a1 :sweat_smile: Thanks for finding that, I didn't realize `secp256k1::Scalar` could be zero. 

Could you please try again now and let me know if `Into<MaybeScalar>` works for tweaks with `secp256k1::Scalar::ZERO`?

Comment (conduition) [reply src/key_agg.rs]: P.s. I've updated `secp` in `musig2`, in this commit: https://github.com/conduition/musig2/commit/a821f6129cee192579ee1a1765dd42e16835cb3d Changelog entry here: https://github.com/conduition/musig2/commit/7937eb2e07f5485dd4b4bbabd10286442bcce35b (you may want to rebase)

Comment (starius) [reply src/key_agg.rs]: Thanks for fixing that panic!
I rebased and squashed it. The patch is now smaller, no `IntoTweak` type.
It works now, no panics.

**Review by conduition — COMMENTED** [review]
Looks good! I have a couple of minor suggestions to apply to improve the unit tests: https://github.com/starius/musig2/pull/2

**Review by conduition — APPROVED** [review]
LGTM

**Comment by conduition** [comment]
Thanks for all your help @starius :100: This change has been deployed alongside [a few others](https://github.com/conduition/musig2/compare/a088ad416e4fc00ee6532d4767b7ffa16aa287e9...5a09b1197b1b5c621a5a9abc60fa95fa84a1da30) as `v0.4.1`.
